source/os/include/threads.h

```diff
@@ -38,40 +39,77 @@ typedef unsigned long threadreturn;
 typedef threadreturn threadcall threadfunc(void*);
 #endif
 
-class atomic_int {
-#ifdef _MSC_VER
-	volatile long value;
-#else
-	volatile int value;
-#endif
+class atomic_int : public std::atomic<int32_t>
+{
 public:
-	int get_basic();
-	void set_basic(int val);
-
-	int operator++();
-	int operator++(int);
-	int operator--();
-	int operator--(int);
-	
-	int operator+=(int value);
-	int operator-=(int value);
-
-	int operator|=(int value);
-	int operator&=(int value);
+	// Offered for backwards compatibility
+	int32_t get_basic()
+	{
+		return load();
+	}
 
-	void operator=(int value);
+	// Offered for backwards compatibility
+	void set_basic(int32_t val)
+	{
+		store(val);
+	}
 
-	int exchange(int value);
-	int compare_exchange_strong(int value, int compareTo);
+	// Offered for backwards compatibility
+	int32_t get() const
+	{
+		return load();
+	}
 
 	//Like compare_exchange_strong, but will not return until the value is exchanged
-	void wait_compare_exchange(int xchg, int compareTo, const int spinCount);
-
-	int get() const { return value; }
-	operator int() const { return value; }
+	void wait_compare_exchange(const int32_t xchg, const int32_t compareToOriginal, const int32_t spinCount) {
```

Comment:
Reimplemented using the std::atomic<>::compare_exchange_strong.

the std::atomic version of compare_exchange_strong has the arguments switched, and returns bool instead of the value currently being stored. 

The first argument is also a reference, and has it's value replaced with the value stored in the atomic object at the time the operation is initiated. Previously this would have been the return value.